### PR TITLE
design: add Platform V2: Txn Management doc

### PR DESCRIPTION
**Pull request by danhhz**
Enable use-case isolation and zero-downtime upgrades in SQL transactions.

Rendered: https://github.com/danhhz/materialize/blob/persist_txn_design/doc/developer/design/20230705_v2_txn_management.md

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/developer/design/20230705_v2_txn_management.md
Comment (benesch) [inline doc/developer/design/20230705_v2_txn_management.md:223]: Say a transaction times out (fails its heartbeat, say) after writing data to the data shard but before updating the remap shard. Who's responsible for clearing out the orphaned data in the data shard? The next transaction? Does each transaction start by inverting all data in the data shard that's not yet committed according to the remap shard?

File: doc/developer/design/20230705_v2_txn_management.md
Comment (benesch) [inline doc/developer/design/20230705_v2_txn_management.md:190]: Any early thoughts on how this will get plumbed into the read path? Will `persist_source` somehow abstract this all away? Will it be a layer on top of `persist_source`?

**Review by benesch — COMMENTED** [review]
Was only able to take a quick skim, but was fun to (finally) spend some time thinking about the remap shard.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (benesch) [reply doc/developer/design/20230705_v2_txn_management.md:223]: > oh i just reread your question and saw that you said "heartbeat". note that atm there are no heartbeats in the protocol. not sure if it'll stick, but i'm initially trying very hard to avoid relying on clocks

Was using "heartbeat" loosely! The situation I was envisioning is when an `envd` that is coordinating a txn fails before it manages to commit the txn. Maybe the node that `envd` was running on gets network partitioned away. Rather than heartbeats, could just be another `envd` coming online and bumping the epoch.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:84]:
```suggestion
- `write_at_ts(ts: u64, updates: Vec<(ShardId, Vec<(Row, i64)>)>) -> bool` which
```

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:202]:
```suggestion
            upper: 1
```

for the example to make sense the upper should be `1`

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:224]: The current timestamp oracle (and also the one in my prototype) has the notion of acquiring a write timestamp and then "closing" it when the write completes. From this, you can derive when there are "open" transactions. Then, in my OCC prototype, I suspend periodic timestamp advancement when noticing this situation.

With this, anyone trying to promptly read things has a bad time. With the alternative (as you mentioned) where we periodically advance the remap timestamp anyone trying to do large-ish read-then-write transactions has a bad time. And, of course, any yolo write-transaction coming in while that slow read-then-write tries to complete is invalidating it (the read-then-write transaction). So I feel like either way we slice it, people trying to do large read-then-writes have a bad time. And this is probably a fine limitation to have. (cc @benesch and I talked about this before)

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:249]: The suspense ... 😅

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:204]: largely nit: the remap shard has the _upper_ for both FromTime and ToTime

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [reply doc/developer/design/20230705_v2_txn_management.md:190]: with Bens (in-flight) fast-persist-peeks idea, that logic would also have to run in those peeker tasks though, right?

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [reply doc/developer/design/20230705_v2_txn_management.md:223]: > iv\. probably most importantly, this is only necessary if we support multi-table write txns. if we're _guaranteed_ that all txns are single table, then we never have to revert anything because any write that makes it to a table data shard is guaranteed to be the whole transaction. if you squint, this is a bit like a special case of how crdb's [parallel commits](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20180324_parallel_commit.md) considers a txn committed if all the intents listed in the txn record went through

Is this true for read-then-write transactions. They would present the same way, you write the updates to the data shard, then you try and "link" those updates into the remap shard. If you notice a conflict then, you have to undo the whole thing and retry the full read-then-write, no? It's because read-then-write implicitly is a potentially-multiple-collections-to-a-single-table transaction.

**Comment by aljoscha** [comment]
Some funky situations that can arise with conflicts/the data-shard-cleanup-protocol (though I think it's all correct iff we do cleanups when noticing a conflict):

Assume two writers `w1` and `w2`, one table, `w1` wants to blind-append `B`, `w2` wants to blind-append `C`.

1.

data/remap shards:
```
data0:
    (A, T0-0, 1)
    upper: 1
remap:
    ((table0, 1), 2023-01-01, 1)
```

2.

Writer `w1` writes `B` to the data shard, so we have:
```
data0:
    (A, T0-0, 1)
    (B, T0-1, 1)
    upper: 2
remap:
    ((table0, 1), 2023-01-01, 1)
```

**Writer `w1` is stalled a bit in trying to "link" things into the remap shard.**

3.

Writer `w2` tries to append `C` to the data shard, at the expected upper `1`. This fails, so it goes to the remap shard to see if it has moved on. It hasn't, so it assumes it needs to retract from the data shard while writing its `C`. Now we have:
```
data0:
    (A, T0-0, 1)
    (B, T0-1, 1)
    (B, T0-2, -1)
    (C, T0-2, 1)
    upper: 3
remap:
    ((table0, 1), 2023-01-01, 1)
```

**Still, no-one has managed to write to the remap shard!**

4.

Writer `w1` managed to write to the remap shard! We now have:
```
data0:
    (A, T0-0, 1)
    (B, T0-1, 1)
    (B, T0-2, -1)
    (C, T0-2, 1)
    upper: 3
remap:
    ((table0, 1), 2023-01-01, 1)
    ((table0, 1), 2023-01-01, -1)
    ((table0, 2), 2023-01-02, 1)
```

**This is correct, and writer `w1`'s write now becomes "visible".

5.

Writer `w2` tries to link the retraction plus its append into the remap shard. It fails, because the remap shard has moved on. It then tries to append it's change the the data shard, at the upper indicated in the remap shard (this is not optimal, and we could be clever, but it seems correct). This fails because the data shard has moved on (by it's own earlier doing, in fact). It then writes a retraction for everything that is beyond the upper that is linked into the remap shard, plus its append. We get:
```
data0:
    (A, T0-0, 1)
    (B, T0-1, 1)
    (B, T0-2, -1)
    (C, T0-2, 1)
    (B, T0-3, 1)
    (C, T0-3, -1)
    (C, T0-3, 1)
    upper: 4
remap:
    ((table0, 1), 2023-01-01, 1)
    ((table0, 1), 2023-01-01, -1)
    ((table0, 2), 2023-01-02, 1)
```

Again: these repeated retractions could be optimized away with smartness, but ... 🤷 . Also, notice that the retraction of a `-1` entry is again a `1`.

6.

Writer `w2` manages to "link" it's writes into the remap shard. Now we have:
```
data0:
    (A, T0-0, 1)
    (B, T0-1, 1)
    (B, T0-2, -1)
    (C, T0-2, 1)
    (B, T0-3, 1)
    (C, T0-3, -1)
    (C, T0-3, 1)
    upper: 4
remap:
    ((table0, 1), 2023-01-01, 1)
    ((table0, 1), 2023-01-01, -1)
    ((table0, 2), 2023-01-02, 1)
    ((table0, 2), 2023-01-02, -1)
    ((table0, 4), 2023-01-02, 1)

```

Now all the writes have become visible!

**Notice how upper `3` of the data shard never became "visible"/linked into the remap shard, because it was intermediate/incorrect state.**

I hope this does make sense to you and is a correct analysis/description of a situation that can happen! Wdyt? An alternative situation would have been that writer `w1` writes to the data shard, but then get's stalled writing to the remap shard. Then writer `w2` writes to both the data shard and remap shard (with a retraction of `w1`'s write). Then, `w1` has to retry the whole thing and also will eventually succeed.

**Comment by aljoscha** [comment]
Some more, loosely related comments around the framing of this. Sorry about these coming out so piece meal... :sweat_smile: 

### Regarding perf/cost

- It might sound bad that a table write now requires 2 persist writes. But that's not so bad compared to the current situation where we do `num_table` writes, and where we also do `num_table` writes every second when idling.
- We are trading off write complexity vs. read. With the remap solution all table reads now become slightly more expensive. But we might get clever and cash the remap shard on all machines.

### Why do we need this for use-case isolation?

The remap-shard idea itself is not what enables scalable table writes. It's how they are used in your example `subscribe`/`write_at_ts`/`read_and_write_ts` methods, which together implement _optimistic concurrency control_ (OCC). This fact is there in the doc but not super obvious.

There are other approaches to distributed txn management but OCC seems to best fit our requirements. You mention those in the non-goals but when we proceed we likely need to spell them out more explicitly.

Raw comments around that:
- Lock-based approaches can help with contention but come with higher base-line overhead. We want to optimize for (potentially large) un-contended, blind writes. -> OCC is a good solution
- It's clear that the current process-level lock doesn't scale.
- We could use OCC with the current bump-every-table-every second approach, but that periodic append conflicts with every write attempt. With the remap-shard approach, a write to an un-contended table only conflicts on the write to the remap shard, so the data written to the data shard can be re-used in a second attempt.
- Then there's other approaches like Calvin, et. al...

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [inline doc/developer/design/20230705_v2_txn_management.md:184]: If you need some design inspiration, the transaction protocol from this paper is fairly similar: http://www.vldb.org/pvldb/vol10/p685-zamanian.pdf/.

They talk a lot about using RDMA, which is not relevant here, but at a very high level it's similar in the sense that they:
1. Write some data down with a timestamp.
2. Mark that data committed by updating a map of data -> committed timestamp.

We read this in the paper reading group and I can't remember if you were there that week. If so, then ignore this.

**Comment by danhhz** [comment]
> Some more, loosely related comments around the framing of this. Sorry about these coming out so piece meal... 😅

not at all! please keep saying things as they come to mind :)

> ### Regarding perf/cost
> * It might sound bad that a table write now requires 2 persist writes. But that's not so bad compared to the current situation where we do `num_table` writes, and where we also do `num_table` writes every second when idling.

agreed! also, I haven't gotten around to writing them up yet, but I think there are some ways to get down from `2 * num_txns + passage_of_time` to `num_txns + passage_of_time` for single-table txns if we decide we care

> * We are trading off write complexity vs. read. With the remap solution all table reads now become slightly more expensive. But we might get clever and cash the remap shard on all machines.

yep. we could also write down the reclocked version of the table data. lots of options here

> ### Why do we need this for use-case isolation?
> The remap-shard idea itself is not what enables scalable table writes. It's how they are used in your example `subscribe`/`write_at_ts`/`read_and_write_ts` methods, which together implement _optimistic concurrency control_ (OCC). This fact is there in the doc but not super obvious.
> 
> There are other approaches to distributed txn management but OCC seems to best fit our requirements. You mention those in the non-goals but when we proceed we likely need to spell them out more explicitly.
> 
> Raw comments around that:
> 
> * Lock-based approaches can help with contention but come with higher base-line overhead. We want to optimize for (potentially large) un-contended, blind writes. -> OCC is a good solution
> * It's clear that the current process-level lock doesn't scale.
> * We could use OCC with the current bump-every-table-every second approach, but that periodic append conflicts with every write attempt. With the remap-shard approach, a write to an un-contended table only conflicts on the write to the remap shard, so the data written to the data shard can be re-used in a second attempt.
> * Then there's other approaches like Calvin, et. al...

agreed that I need to call this out more explicitly. current version of the doc is meant to be enough to help frame discussion with people that already have the context loaded up :). in my mind there's two problems being solved here: 1) is make tables distributed _at all_ and 2) is make them _isolated from each other_. even if we didn't want to go with OCC, we'd still have to do work to the current in-process write lock to make it distributed. in some sense, i've been focusing so far more on (1), where remap is replacing the lock, than (2), which is more the domain of OCC vs PCC. we could just as easily build a PCC protocol on top of the remap shard

> Some funky situations that can arise with conflicts/the data-shard-cleanup-protocol (though I think it's all correct iff we do cleanups when noticing a conflict):
> 
> Assume two writers `w1` and `w2`, one table, `w1` wants to blind-append `B`, `w2` wants to blind-append `C`.
> 
> 
> data/remap shards:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     upper: 1
> remap:
>     ((table0, 1), 2023-01-01, 1)
> ```
> 
> 
> Writer `w1` writes `B` to the data shard, so we have:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     (B, T0-1, 1)
>     upper: 2
> remap:
>     ((table0, 1), 2023-01-01, 1)
> ```
> 
> **Writer `w1` is stalled a bit in trying to "link" things into the remap shard.**
> 
> 
> Writer `w2` tries to append `C` to the data shard, at the expected upper `1`. This fails, so it goes to the remap shard to see if it has moved on. It hasn't, so it assumes it needs to retract from the data shard while writing its `C`. Now we have:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     (B, T0-1, 1)
>     (B, T0-2, -1)
>     (C, T0-2, 1)
>     upper: 3
> remap:
>     ((table0, 1), 2023-01-01, 1)
> ```
> 
> **Still, no-one has managed to write to the remap shard!**
> 
> 
> Writer `w1` managed to write to the remap shard! We now have:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     (B, T0-1, 1)
>     (B, T0-2, -1)
>     (C, T0-2, 1)
>     upper: 3
> remap:
>     ((table0, 1), 2023-01-01, 1)
>     ((table0, 1), 2023-01-01, -1)
>     ((table0, 2), 2023-01-02, 1)
> ```
> 
> **This is correct, and writer `w1`'s write now becomes "visible".
> 
> 
> Writer `w2` tries to link the retraction plus its append into the remap shard. It fails, because the remap shard has moved on. It then tries to append it's change the the data shard, at the upper indicated in the remap shard (this is not optimal, and we could be clever, but it seems correct). This fails because the data shard has moved on (by it's own earlier doing, in fact). It then writes a retraction for everything that is beyond the upper that is linked into the remap shard, plus its append. We get:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     (B, T0-1, 1)
>     (B, T0-2, -1)
>     (C, T0-2, 1)
>     (B, T0-3, 1)
>     (C, T0-3, -1)
>     (C, T0-3, 1)
>     upper: 4
> remap:
>     ((table0, 1), 2023-01-01, 1)
>     ((table0, 1), 2023-01-01, -1)
>     ((table0, 2), 2023-01-02, 1)
> ```
> 
> Again: these repeated retractions could be optimized away with smartness, but ... 🤷 . Also, notice that the retraction of a `-1` entry is again a `1`.
> 
> 
> Writer `w2` manages to "link" it's writes into the remap shard. Now we have:
> 
> ```
> data0:
>     (A, T0-0, 1)
>     (B, T0-1, 1)
>     (B, T0-2, -1)
>     (C, T0-2, 1)
>     (B, T0-3, 1)
>     (C, T0-3, -1)
>     (C, T0-3, 1)
>     upper: 4
> remap:
>     ((table0, 1), 2023-01-01, 1)
>     ((table0, 1), 2023-01-01, -1)
>     ((table0, 2), 2023-01-02, 1)
>     ((table0, 2), 2023-01-02, -1)
>     ((table0, 4), 2023-01-02, 1)
> ```
> 
> Now all the writes have become visible!
> 
> **Notice how upper `3` of the data shard never became "visible"/linked into the remap shard, because it was intermediate/incorrect state.**
> 
> I hope this does make sense to you and is a correct analysis/description of a situation that can happen! Wdyt? An alternative situation would have been that writer `w1` writes to the data shard, but then get's stalled writing to the remap shard. Then writer `w2` writes to both the data shard and remap shard (with a retraction of `w1`'s write). Then, `w1` has to retry the whole thing and also will eventually succeed.
> 
> edit: things are only this funky when there's contention. For the normal case things should be smooth for the cases we want to optimize for.

definitely something I'm meaning to spell out more explicitly but yeah this is exactly how the multi-table rollback protocol would work. i mentioned this in some of the comments i'm leaving now, but if we restrict ourselves to single-table txns, we can have a more advantageous "roll forward" protocol

i'm not sure, was there a question here or just checking your understanding? :)

**Comment by aljoscha** [comment]
> i'm not sure, was there a question here or just checking your understanding? :)

the latter (checking my understanding) 😅

File: doc/developer/design/20230705_v2_txn_management.md
Comment (mjibson) [inline doc/developer/design/20230705_v2_txn_management.md:43]: This should maybe get replaced with "highly optimized webhooks", or at the least it should be added at a higher priority than tables.

**Comment by danhhz** [comment]
@aljoscha @jkosh44 Updated this to reflect the latest thinking, so it should be ready to go.

@MaterializeInc/qa Would love to hear your thoughts about testing. This is primarily motived by a bunch of followup Pv2 work that depends on it, but it does happen to have a side benefit of isolating us from large numbers of tables. (Philip will also be happy to hear that it will help some with the high CRDB usage.)

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [inline doc/developer/design/20230705_v2_txn_management.md:288]: > As well as the set of registered shards?

Ditto

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [inline doc/developer/design/20230705_v2_txn_management.md:328]: What if the `as_of` is past the latest non-empty write? For example:

- Shard A's latest non-empty write is `10`.
- The txn shard's latest non-empty write is `20`.
- Someone requests a snapshot of shard A as of `30`?

I think the correct thing to do here is to use a timestamp of `30`, not `20` which is what I think this is suggesting?

As a follow-up, if we do choose a timestamp of `30` and shard A never gets updated again, but the txn shard eventually advances past `30`, then how do we find this out and complete the read?

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [inline doc/developer/design/20230705_v2_txn_management.md:300]: Can these retractions cause other transactions to abort due to conflicting timestamps?

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [inline doc/developer/design/20230705_v2_txn_management.md:146]: I don't think you need to include the details here, but just for my own understanding, in order to remove a data shard do you need to advance it's physical upper to match the logical upper from the transaction shard?

Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:146]: yeah, you would

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:221]: done

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:288]: done

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:300]: Yeah, you've correctly anticipated an issue I found the hard way in the WIP integration branch. Even more than aborts, it seems everything gets real panic-y because of the single-envd assumptions.

I think the fix actually works out pretty neatly. We split what is currently "apply" into two steps: "apply" (which only does the compare_and_append_batch and which is the bit to unblock reads) and "tidy" (which retracts the completed apply work). The tidy step then can take a timestamp, which lets us involve the timestamp oracle (though I think in practice we'll want to just sneak it in with the next txn commit to save both the timestamp and the crdb write).

Updated the doc to reflect the above. The "tidy" jargon is a strawman, taking counter-proposals if you have them.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:328]: I'm a bit confused why the last non-empty write at `20` is relevant, but if I'm following the question correctly, it's actually suggesting `10`.

If the last non-empty write to some data shard is at `10`, and the txn shard's upper is past some snapshot as_of `X`, then we know that `[10,X]` is empty in A, so it's safe to just do a snapshot directly on the shard at `10`. That's true even if we haven't yet finished applying everything in A: e.g. if the write at `10` hasn't applied yet then the snapshot will just block until it does.

Otoh, if the txn shard's upper is not past `X` yet, then anything could happen and we have to wait until it does before saying anything about the snapshot (which is normally what happens with snapshots).

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [reply doc/developer/design/20230705_v2_txn_management.md]: There's are all mostly synonyms, but in case you like one better:

- cleanup
- prune
- trim

PostgreSQL refers to a similar process as cleanup (https://www.postgresql.org/docs/current/runtime-config-wal.html#GUC-ARCHIVE-CLEANUP-COMMAND).

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [reply doc/developer/design/20230705_v2_txn_management.md:328]: > I'm a bit confused why the last non-empty write at `20` is relevant, but if I'm following the question correctly, it's actually suggesting `10`.

> Otoh, if the txn shard's upper is not past `X` yet, then anything could happen and we have to wait until it does before saying anything about the snapshot

I think I'm probably confused as to how exactly it's supposed to work or asked the question incorrectly. If `X = 30` I thought that saying the last non-empty write was at `20` implied that the txn shard's upper was not past `X`.

Are the data shard uppers always equal to the last applied non-empty write on that shard? And is the txn-shard upper always greater than or equal to the last non-empty write of all shards (potentially greater than due to registrations and retractions)? 


Let me try and walk through the example in more detail:

1. Register shard `A` at ts `1`.
2. Register shard `B` at ts `2`.
3. Commit and apply write on shard `A` at ts `3`.
4. Commit and apply write on shard `B` at ts `4`.
5. Initiate snapshot of shard `A` at ts `10`.
  a. The timestamp does not get translated and stays at `10` because the txn shard upper is `5` which is not past `10` (I think).
  b. The snapshot is blocked because we have to wait until the txn shard upper advances past `10`.
6. Commit and apply write on shard `B` at ts `20`.
  a. Moving the txn shard upper to `21`.

My questions are: 

- Was I correct in saying the the timestamp does not get translated and stays at `10`? 
- If so, what mechanism informs the blocked snapshot that the logical upper has advanced and the snapshot can complete?

Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:328]: (whoops initially typed that wrong, ignore the version of this comment in your email 😅)

Comment (jkosh44) [reply doc/developer/design/20230705_v2_txn_management.md:328]: Ok, that clears everything up. I think what I was missing was that before we translate the timestamp we must first wait for the txn shard's upper to pass our timestamp (i.e. call `TxnsRead::update_gt(X)`).

Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md]: Cool! Called these three steps out explicitly in the doc

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:72]: If we want to be very accurate: The oracle and the write lock/mutex are separate but they interact. The lock "protects" writes, not access to the timestamp oracle. A read-then-write transaction does:

1. acquire write lock (this prevents other writes from going through, which yes, indirectly also means that the oracle read timestamp will not advance)
2. get a read timestamp from the oracle
3. perform read
4. perform write (which also advances the read timestamp that the oracle keeps)
5. release write lock

Or I don't understand what _surrounding_ means here. 😅

File: doc/developer/design/20230705_v2_txn_management.md
Comment (aljoscha) [inline doc/developer/design/20230705_v2_txn_management.md:177]: Are they linearized or serialized through the txns shard? IMO, the txn logic gives us serializability and the way we pick timestamps ensures linearizability. Which together ensures we provide _strict serializability_.

**Review by aljoscha — APPROVED** [review]
Still a lovely design and a nice doc! Plus I like how iterating on it with Joe has clarified things.

I did leave two comments that are quite in the weeds and shouldn't block merging.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [reply doc/developer/design/20230705_v2_txn_management.md:72]: > which yes, indirectly also means that the oracle read timestamp will not advance

If I'm remembering correctly I don't think this is 100% accurate. The read timestamp will still advance from writes to system tables or from periodic table advancements, which both don't require the global write lock.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (jkosh44) [reply doc/developer/design/20230705_v2_txn_management.md:177]: > IMO, the txn logic gives us serializability and the way we pick timestamps ensures linearizability.

I think technically all write only txns are linearized through the txns shard. If a write `B` starts after write `A` completes then it's guaranteed to have a larger timestamp, because of the upper advancing. I.e. the total order of writes matches the real time ordering.

If we're talking about read then write txns, then I don't think the txn logic even provides serializability, at best it provides read committed and monotonic writes. The timestamp selection is what ensures both serializability and linearizability.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:72]: Updated this with two bullets provided offline by joe.

File: doc/developer/design/20230705_v2_txn_management.md
Comment (danhhz) [reply doc/developer/design/20230705_v2_txn_management.md:177]: ☝️ I had a response staged, but joe's is much better, so I deleted it :)

Change this to "All txn writes are linearized"

**Review by danhhz — COMMENTED** [review]
TFTRs!!

**Pull request merged**
